lib/src/fixed_datetime_codec.dart

```diff
@@ -0,0 +1,153 @@
+/// * C	a digit used in the time scale component “century”
+/// * h	a digit used in the time scale component “clock hour”
+/// * m	a digit used in the time scale component “clock minute”
+/// * s	a digit used in the time scale component “clock second”
+/// as specified in the ISO 8601 standard.
+///
+/// Non-allowed characters in the format [_pattern] are ignored, therefore
+/// `YYYY kiwi MM` is the same format string as `YYYY------MM`.
+///
+/// Note: this class differs from [DateFormat] in that here, the characters are
+/// treated literally, i.e. the format string `YYY` matching `996` would result in
+/// the same as calling `DateTime(996)`. [DateFormat] on the other hand uses the
+/// specification in https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table,
+/// the format string (or "skeleton") `YYY` specifies only the padding, so
+/// `1996` would be a valid match. This limits is use to format strings
+/// containing delimiters, as the parser would not know how many digits to take
+/// otherwise.
+///
+/// Also, this parser does not know about locales and parses to the current
+/// locale only.
+class FixedDateTimeCodec {
+  static const _validChars = ['Y', 'M', 'D', 'E', 'C', 'h', 'm', 's'];
+
+  final String _pattern;
+  final _occurences = <String, _Range>{};
+
+  FixedDateTimeCodec(this._pattern) {
+    String current = '';
+    for (var i = 0; i < _pattern.length; i++) {
+      var char = _pattern[i];
+      if (!_validChars.contains(char)) {
+        current = '';
+        continue;
+      }
+      var newChar = current != char;
+      if (newChar) {
+        var hasNotSeenBefore = !_occurences.containsKey(char);
+        if (hasNotSeenBefore) {
+          _occurences[char] = _Range(i, i);
+        } else {
+          throw Exception(
+              'The _pattern string "$_pattern" contains multiple instances of the formatting char block $char, both at position ${_occurences[char]!.from} and $i');
+        }
+      }
+      current = char;
+      _occurences.update(current, (value) => _Range(value.from, value.to + 1));
+    }
+  }
+
+  /// Convert a datetime to a string exactly as specified by the [_pattern].
+  String encode(DateTime dt) {
+    var startingPoints =
+        _occurences.map((key, value) => MapEntry(value.from, key));
+    var sb = StringBuffer();
+    for (var i = 0; i < _pattern.length; i++) {
+      if (startingPoints.containsKey(i)) {
+        var key = startingPoints[i];
+        var length = _occurences[key]!.length;
+        var number = _extractNumFromDateTime(key, dt);
+        var numAsString = number.toString();
+        if (numAsString.length > length) {
+          throw Exception(
+              "The datetime $dt cannot be parsed as $number is longer than the _pattern $key allows");
+        } else {
+          sb.write(numAsString.padLeft(length, '0'));
+        }
+        i += length - 1;
+      } else {
+        sb.write(_pattern[i]);
```

Comment:
This could probably be optimized to jump to the next `startingPoint`, but I didn't manage to do that without making the code a lot more complicated. Should I still try to do so?